site/src/hooks/useClipboard.ts

```diff
@@ -1,44 +1,178 @@
-      await window.navigator.clipboard.writeText(text);
-      setIsCopied(true);
-      window.setTimeout(() => {
-        setIsCopied(false);
-      }, 1000);
-    } catch (err) {
-      const input = document.createElement("input");
-      input.value = text;
-      document.body.appendChild(input);
-      input.focus();
-      input.select();
-      const result = document.execCommand("copy");
-      document.body.removeChild(input);
-      if (result) {
-        setIsCopied(true);
-        window.setTimeout(() => {
-          setIsCopied(false);
-        }, 1000);
-      } else {
-        const wrappedErr = new Error(
-          "copyToClipboard: failed to copy text to clipboard",
-        );
-        if (err instanceof Error) {
-          wrappedErr.stack = err.stack;
-        }
-        console.error(wrappedErr);
-      }
+import { useCallback, useEffect, useState } from "react";
+import { useEffectEvent } from "./hookPolyfills";
+
+type UseClipboardResult = Readonly<{
+  isCopied: boolean;
+  copyToClipboard: () => Promise<void>;
+}>;
+
+export const useClipboard = (textToCopy: string): UseClipboardResult => {
+  // Can't initialize clipboardText with a more specific value because reading
+  // is an async operation
+  const [clipboardText, setClipboardText] = useState("");
+
+  // Copy events have a ClipboardEvent associated with them, but sadly, the
+  // event only gives you information about what caused the event, not the new
+  // data that's just been copied. Have to use same handler for all operations
+  const syncClipboardToState = useEffectEvent(async () => {
+    const result = await readFromClipboard();
+    setClipboardText((current) => (result.success ? result.value : current));
+  });
+
+  useEffect(() => {
+    // Focus event handles case where user navigates to a different tab, copies
+    // new text, and then comes back to Coder
+    window.addEventListener("focus", syncClipboardToState);
+    window.addEventListener("copy", syncClipboardToState);
+    void syncClipboardToState();
+
+    return () => {
+      window.removeEventListener("focus", syncClipboardToState);
+      window.removeEventListener("copy", syncClipboardToState);
+    };
+  }, [syncClipboardToState]);
+
+  const copyToClipboard = useCallback(async () => {
+    const result = await writeToClipboard(textToCopy);
+    if (result.success) {
+      void syncClipboardToState();
+      return;
     }
-  };
+  }, [syncClipboardToState, textToCopy]);
 
   return {
-    isCopied,
-    copy,
+    copyToClipboard,
+    isCopied: textToCopy === clipboardText,
   };
 };
+
+type VoidResult = Readonly<
+  { success: true; error: null } | { success: false; error: Error }
+>;
+
+type ResultWithData<T = unknown> = Readonly<
+  | { success: true; value: T; error: null }
+  | { success: false; value: null; error: Error }
+>;
+
+type Result<T = unknown> = void extends T ? VoidResult : ResultWithData<T>;
+
+async function readFromClipboard(): Promise<Result<string>> {
```

Comment:
I did not understand this part. Why do we need to read something from the clipboard? I would expect to just write things in it.

Reply:
Maybe listing the flow using bullet points would make things easier to understand for me.

Reply:
It's mainly for setting up a way to watch the clipboard, since there isn't an observer-based API for it, like there is for a lot of other browser APIs

The main idea is that `isCopied` is always a derived value that comes from checking whether the text passed into the hook matches the text currently in the clipboard. And the only way to check the current clipboard is by reading from it – if we only write to it, then we can't make any assumptions about what's actually in the clipboard, because it can also change from sources outside React

So basically, there's three main "sync cues":
1. The user copies our text to the clipboard
2. The user changes their clipboard text while in the same page
3. The user changes to a different tab, and then comes back

In all of the cases, we don't have a direct, event-based way to check the clipboard contents to verify that the content matches. You would think that writing to the clipboard would expose that information, but the event objects don't actually contain the new clipboard text. So in all three cases, the only choice is reading from it

Reply:
Why do we need to check if the content in the clipboard matches the one passed into the hook? Is there a use case in the app where this can happen?